### PR TITLE
database: Faster external service deletion

### PR DESCRIPTION
**Pull request by ryanslade**
Instead of scanning the entire repo table looking for orphaned repos
after an external service has been deleted, we instead only scan repos
we know may have been affected.

We had to use a temporary table since if the deletion was done in a CTE
it is not visible in subsequent statements.

We also remove the trigger after external service is deleted as we 
take care of it in code.

Closes: https://github.com/sourcegraph/sourcegraph/issues/19513

### CODE REVIEW

**Comment by ryanslade** [comment]
In our testing it was almost instant.

**Review by tsenart — APPROVED** [review]
Nice! I wonder what happened with that function I wrote though? Was it completely broken? :D

**Comment by tartarza112244** [comment]
+1 Approve

**Review by tartarza112244 — APPROVED** [review]

**Review by tartarza112244 — APPROVED** [review]

**Comment by ryanslade** [comment]
> Nice! I wonder what happened with that function I wrote though? Was it completely broken? :D

It's fine but we wanted to keep things in code and avoid the hack if disabling triggers.

**Comment by ryanslade** [comment]
@tsenart @asdine  We have a potential issue since deleting a repo by setting `deleted_at` and then upserting is no longer supported.. We don't seem to do that anywhere in the code but it could cause inconsistencies in future.

We could change the `Delete` function to take multiple external services and call it at the end of upsert but I'm not sure how that'll affect performance.

**Comment by tsenart** [comment]
> We have a potential issue since deleting a repo by setting deleted_at and then upserting is no longer supported.. We don't seem to do that anywhere in the code but it could cause inconsistencies in future.

We should probably add a check that returns an error in Upsert if we try to delete through it.

> We could change the Delete function to take multiple external services and call it at the end of upsert but I'm not sure how that'll affect performance.

This makes sense to me. Can you elaborate on the performance concerns?

**Review by unknwon — APPROVED** [review]

File: internal/database/external_services.go [hunk 1]
```diff
@@ -879,7 +879,52 @@ func (e *ExternalServiceStore) Delete(ctx context.Context, id int64) (err error)
...
+    repo_id int
+) ON COMMIT DROP`)); err != nil {
+		return errors.Wrap(err, "creating temporary table")
+	}
+
+	// Delete external service <-> repo relationships, storing the affected repos
+	if err := tx.Exec(ctx, sqlf.Sprintf(`
+	WITH deleted AS (
+	   DELETE FROM external_service_repos
+	       WHERE external_service_id = %s
```
Comment (ryanslade) [inline internal/database/external_services.go:896]: > We could change the Delete function to take multiple external services and call it at the end of upsert but I'm not sure how that'll affect performance.

@tomas I'm worried about what'll happen if we change this to an `in` or `exists` clause. I seem to remember we did this in the past and it no longer used the index. 

I'm happy to merge the "slow" version anyway and we can do another round of optimisation if we need to.

Comment (tomas) [reply internal/database/external_services.go:896]: I have nothing to do with this. Why am I being mentioned??

Comment (ryanslade) [reply internal/database/external_services.go:896]: My bad, was meant to tag @tsenart 😄

Comment (tsenart) [reply internal/database/external_services.go:896]: > I'm worried about what'll happen if we change this to an in or exists clause

We should just test it.

Comment (ryanslade) [reply internal/database/external_services.go:896]: I've had a look this morning and it's quite messy. We return custom error type, `externalServiceNotFoundError` which assumes only external service. We could update that to have a slice of external services but what do we return if only a subset of external services can't be found.

I think given that we don't currently use upsert to delete we can leave the code as is and I propose that we deprecate soft deletion of external services. Unless I'm missing some context it only appears to complicate our code.

I'll ask in Slack if anyone knows a good reason why we need to undelete an external service.

Comment (ryanslade) [reply internal/database/external_services.go:912]: As an extra safety net I just pushed a change that returns an error if you try and delete an external service via `Upsert`

**Pull request merged**